### PR TITLE
Clarify our OOP semantics across the codebase

### PR DESCRIPTION
**Pull request by NiLuJe**
Basically:

* Use `extend` for class definitions
* Use `new` for object instantiations

That includes some minor code cleanups along the way:

* Updated `Widget`'s docs to make the semantics clearer.
* Removed `should_restrict_JIT` (it's been dead code since https://github.com/koreader/android-luajit-launcher/pull/283)
* Minor refactoring of LuaSettings/LuaData/LuaDefaults/DocSettings to behave (mostly, they are instantiated via `open` instead of `new`) like everything else and handle inheritance properly (i.e., DocSettings is now a proper LuaSettings subclass).
* Default to `WidgetContainer` instead of `InputContainer` for stuff that doesn't actually setup key/gesture events.
* Ditto for explicit `*Listener` only classes, make sure they're based on `EventListener` instead of something uselessly fancier.
* Unless absolutely necessary, do not store references in class objects, ever; only values. Instead, always store references in instances, to avoid both sneaky inheritance issues, and sneaky GC pinning of stale references.
  * ReaderUI: Fix one such issue with its `active_widgets` array, with critical implications, as it essentially pinned *all* of ReaderUI's modules, including their reference to the `Document` instance (i.e., that was a big-ass leak).
* Terminal: Make sure the shell is killed on plugin teardown.
* InputText: Fix Home/End/Del physical keys to behave sensibly.
* InputContainer/WidgetContainer: If necessary, compute self.dimen at paintTo time (previously, only InputContainers did, which might have had something to do with random widgets unconcerned about input using it as a baseclass instead of WidgetContainer...).
* OverlapGroup: Compute self.dimen at *init* time, because for some reason it needs to do that, but do it directly in OverlapGroup instead of going through a weird WidgetContainer method that it was the sole user of.
* ReaderCropping: Under no circumstances should a Document instance member (here, self.bbox) risk being `nil`ed!
* Kobo: Minor code cleanups.

(Lots of files touched, but mostly trivial cosmetic tweaks, most actual code changes are limited to LuaSettings & friends).

Pairs with https://github.com/koreader/koreader-base/pull/1525

Sponsored by `rg '^local [[:alnum:]_\-]+ = [[:alnum:]_\-]+:new'` & `rg "^[[:blank:]]{4}[^.[:blank:]]+[[:blank:]]*=[[:blank:]]*\{\}"` ;p.

- - -
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/9586)

### CODE REVIEW

**Comment by poire-z** [comment]
Looks a bit scary :) v2022.10 or v2022.11 ?

> We very, very, very, very rarely actually feed an existing Widget object
to extend (because that way lies inheritance madness).
In 99% of the cases, we pass it the class object of our new subclass

What is the 1% ?
There must probably be 1 or 2 odd usages that may break if blindly moved from :new to :extend. But I guess you may have found them ? :)

Thank you for having removed all these noisy `assert(self ~= nil)` (added by a dev who didn't trust himself very much :)

**Comment by NiLuJe** [comment]
> What is the 1%

I... don't think we actually do it anywhere, but it obviously still worts, because a Widget is still a table, after all ;).
(This was mostly a documentation clarification).

**Comment by NiLuJe** [comment]
> v2022.10 or v2022.11 ?

Dealer's choice ;).

(I've been running with it applied for a couple days and nothing's blown up so far, but I perfectly understand if we want to wait a bit ;)).

**Comment by NiLuJe** [comment]
> There must probably be 1 or 2 odd usages that may break if blindly moved from :new to :extend. But I guess you may have found them ? :)

AFAICT, everything properly instantiates w/ new, so we're good.

(Also, in quite a few of the classes involved, new actually == extend; c.f., CacheItem for example).

File: frontend/ui/widget/widget.lua [hunk 1]
```diff
@@ -14,44 +14,42 @@ local EventListener = require("ui/widget/eventlistener")
...
+It only setups the metatable (or prototype chain) and will not initiate a real instance, i.e. call self:init().
 
-@tparam Widget baseclass
+@tparam table subclass
 @treturn Widget
 ]]
-function Widget:extend(baseclass)
-    local o = baseclass or {}
+function Widget:extend(subclass)
+    local o = subclass or {}
```
Comment (poire-z) [inline frontend/ui/widget/widget.lua:27]: (The name `baseclass` was wrong (didn't ever noticed).
But `subclass` although more correct, feels yet a bit wrong : we pass it a table/prototype (with default properties) and it is returned as being upgraded to be a subclass.
But ok, we pass what we want to become a subclass.)

Comment (NiLuJe) [reply frontend/ui/widget/widget.lua:27]: Yeah, I didn't think of anything better ;o).

(I'm not necessarily super fond of "extend" either, FWIW).

Comment (NiLuJe) [reply frontend/ui/widget/widget.lua:27]: But I kinda like `prototype`, so... maybe `subclass_prototype` would work better?

**Comment by Frenzie** [comment]
Don't have internet yet, some issue with the fiber in my street, so might as well throw it in.

**Review by NiLuJe — COMMENTED** [review]
> Reviewed 201 of 202 files at r1, 1 of 1 files at r2, 10 of 10 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/koreader/koreader/9586)* status: all files reviewed, 1 unresolved discussion (waiting on @Frenzie, @pazos, and @poire-z)

**Comment by NiLuJe** [comment]
I sure would like to know what's got CircleCI's panties in a bunch, but it's been throwing errors 500 for a few days over here...

**Review by NiLuJe — COMMENTED** [review]
> Reviewed 35 of 37 files at r4, 4 of 4 files at r5, 3 of 3 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/koreader/koreader/9586)* status: all files reviewed, 1 unresolved discussion (waiting on @Frenzie, @pazos, and @poire-z)

File: frontend/device/kobo/device.lua [hunk 2]
```diff
@@ -1235,12 +1235,14 @@ function Kobo:toggleChargingLED(toggle)
...
     --       (when it does, it's an option in the Energy saving settings),
     --       which is why we also limit ourselves to "true" on devices where this was tested.
+    --       FWIW, on older devices, the knob is at "/sys/devices/platform/pmic_light.1/lit".
     -- c.f., drivers/misc/ntx_misc_light.c
     local f = io.open("/sys/devices/platform/ntx_led/lit", "we")
     if not f then
         logger.err("cannot open /sys/devices/platform/ntx_led/lit for writing!")
         return false
     end
+    C.setlinebuf(f)
```
Comment (zwim) [inline frontend/device/kobo/device.lua:1248]: Why is setlinebuf only here?
I would set it globally (or disable it at the end of this method ).

Comment (NiLuJe) [reply frontend/device/kobo/device.lua:1248]: It's per (open) FILE handle ;).

Comment (zwim) [reply frontend/device/kobo/device.lua:1248]: Oh, of course, you are right.

**Comment by poire-z** [comment]
```
InputText: Fix Home/End/Del to behave like a real keyboard
e.g., Del doesn't do weird stuff anymore, and Home/End move across the *current* line.
```

You had me worried - but that's indeed just the Home and End physical keyboard keys.
The  virtual keyboard's image are actually Top and Bottom keys, and use the unchanged scrollToTop/Bottom methods.
So, fine by me :)

**Comment by poire-z** [comment]
>  InputContainer: That... wasn't a deep copy, like, at all.
So, don't do it, because I don't think we need a deep copy at all her,e
seeing as this is already an instance member, not a class one.

Looked, and I agree (in case you need confirmation :)
Just curious: you go at all that trusting your (good) sense ? Or do you look at git history/blame, to see who did the change when, if there was a reason or not, if the person who did the change feel like it had experience or was a newcommer ? 
There's many things I've seen in some early/base widgets that looked odd (some stuff I remembed related to align= stuff), that I didn't dare changing because "if it works don't fix it" :)

**Comment by NiLuJe** [comment]
> Just curious: you go at all that trusting your (good) sense ? Or do you look at git history/blame, to see who did the change when, if there was a reason or not, if the person who did the change feel like it had experience or was a newcommer ?
> There's many things I've seen in some early/base widgets that looked odd (some stuff I remembed related to align= stuff), that I didn't dare changing because "if it works don't fix it" :)

All of the above? ;o)

(And I'm trying to keep to stuff mostly related to this PR. e.g., parts of the dimen stuff was an old, old workaround... because of base classes being registered via new, so with init doing funky stuff to base classes ;)).

Let's say I'm cautiously optimistic it won't break too many things ;p.

**Review by NiLuJe — COMMENTED** [review]
> Reviewed 3 of 7 files at r7, 3 of 4 files at r8, 3 of 3 files at r9, 5 of 8 files at r10, 27 of 27 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/koreader/koreader/9586)* status: all files reviewed, 2 unresolved discussions (waiting on @Frenzie, @pazos, @poire-z, and @zwim)

**Review by NiLuJe — COMMENTED** [review]
> Reviewed 3 of 3 files at r12, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/koreader/koreader/9586)* status: all files reviewed, 2 unresolved discussions (waiting on @Frenzie, @pazos, @poire-z, and @zwim)

**Comment by poire-z** [comment]
Crash when Wikipedia lookup:
```
10/08/22-16:24:44 WARN  error in wrapped function: frontend/apps/reader/modules/readerdictionary.lua:970: attempt to get length of field 'dict_window_list' (a nil value)
stack traceback:
 frontend/apps/reader/modules/readerdictionary.lua:970: in function 'showDict'
 frontend/apps/reader/modules/readerwikipedia.lua:536: in function 'lookupWikipedia'
 frontend/apps/reader/modules/readerwikipedia.lua:392: in function <frontend/apps/reader/modules/readerwikipedia.lua:391>
 [C]: in function 'xpcall'
```
Dunno how you'll want to fix that : :new ? re-set all the ReaderDict fields in ReaderWikipedia ? Call super/ReaderDict:init() ?

**Comment by NiLuJe** [comment]
@poire-z: I was a bit unclear on what exactly the `dict_window_list` array's *raison d'être* was, so it will kind of depend on that?

**Comment by NiLuJe** [comment]
Okay, yeah, looking at DictQuickLookup, it needs to be a static class member (if only for the "close all open windows on hold" feature), I'll fix that ;).

**Comment by poire-z** [comment]
> InputText: Fix Home/End/Del physical keys to behave sensibly.

Works when a scrollable has not yet scrolled. Once you scroll and the first line is not the top line, all go messy :) Home moves some lines up, End too.
I guess it's because your added:
`current_line_num = 1, -- index of the currently edited line (i.e., where our cursor is)`
is actually the index of the currently edited line among the visible lines, and not related to the top (undisplayed) line.
So, your recent fix moves it to not the right place:
```lua
function TextBoxWidget:moveCursorHome()
    self:moveCursorToCharPos(self.vertical_string_list[self.current_line_num] and self.vertical_string_list[self.current_line_num].offset or #self.charlist)
end
```

Looking at fixing that.

**Comment by poire-z** [comment]
^ will PR that fix, there are a few minor other related cleanups needed.

**Pull request merged**